enhancements/machine-api/machine-health-check-pause-on-upgrade.md

```diff
@@ -0,0 +1,188 @@
+
+## Motivation
+
+During cluster upgrades nodes get temporarily unhealthy, which might trigger remediation when a machineHealthCheck
+resource is targeting those nodes, and they don't get healthy before the configured timeout. This remediation delays
+the cluster upgrade significantly especially on bare metal clusters, and causes needless interruption of customer
+workloads while they are transferred to other nodes.
+
+### Goals
+
+- Automatically pause all machineHealthChecks during cluster upgrades
+- Automatically unpause all machineHealthChecks when cluster upgrade finished
+- Get a simple solution done in time for OCP 4.9, by using what we have upstream already (the pause annotation), for
+  avoiding worst case behaviour during cluster upgrades. Look for better solutions for future releases
+
+### Non-Goals
+
+- Track the status of each single node (upgrading or not upgrading) for more fine-grained pausing of remediation
+
+## Proposal
+
+Introduce a new controller, which watches clusterVersion and machineHealthCheck resources, and pauses all
+machineHealthChecks when the clusterVersion status indicates an ongoing cluster upgrade. The controller will unpause the
+machineHealthChecks when the clusterVersion status indicates that the upgrade process finished.
+
+### User Stories
+
+- As a cluster admin I don't want unneeded remediation/reboots of nodes during cluster upgrades
+- As a cluster admin I don't want to manually pause machineHealthChecks during cluster upgrades
+- As a user I don't want my workloads to be transferred to new nodes more often than needed during cluster upgrades
+
+### Implementation Details/Notes/Constraints [optional]
+
+The clusterVersion object has several conditions in its status. Example:
+
+```yaml
+  status:
+    conditions:
+    - lastTransitionTime: "2021-07-05T20:53:02Z"
+      message: Done applying 4.9.0-0.nightly-2021-07-05-092650
+      status: "True"
+      type: Available
+    - lastTransitionTime: "2021-07-06T12:03:02Z"
+      status: "False"
+      type: Failing
+    - lastTransitionTime: "2021-07-05T20:53:02Z"
+      message: Cluster version is 4.9.0-0.nightly-2021-07-05-092650
+      status: "False"
+      type: Progressing
+    - lastTransitionTime: "2021-07-05T20:05:15Z"
+      message: 'Unable to retrieve available updates: currently reconciling cluster
+        version 4.9.0-0.nightly-2021-07-05-092650 not found in the "stable-4.8" channel'
+      reason: VersionNotFound
+      status: "False"
+      type: RetrievedUpdates
+    ...
+```
+
+The machineHealthCheck supports pausing by setting an annotation on it:
+
+```yaml
+apiVersion: machine.openshift.io/v1beta1
+kind: MachineHealthCheck
+metadata:
+  name: example
+  namespace: openshift-machine-api
+  annotations:
+    cluster.x-k8s.io/paused: ""
+  ...
+```
+
+The new controller will add the pause annotation on the machineHealthChecks when the clusterVersion's `Progressing`
```

Comment:
This functionality should be opt-in, we don't want to drastically change default behavior like this.  If users are having a problem with MHC too eagerly deleting instances during upgrades, they can opt-in (or adjust MHC, or pause it themselves).  Making this behavior the default is not a good fit IMO.

Since we don't know who set the annotation on the MHC, we'll need another small annotation on the MHC to indicate to ourselves that we paused it for upgrade and the annotation didn't already exist.